alphagov/govuk_publishing_components
# Remove dependency on .js.erb and puppet for loading our analytics

Hi @andysellick / @JamesCGDS - would you be able to approve this if all is OK? Thanks :+1:

## What

This change creates a JavaScript file which will load in our analytics variables depending on the current domain that the user is on.  I have created `load-analytics.js`.  This contains a `loadUa()` and `loadGa4()` function (they're separate due to a suggestion from @andysellick  and it makes sense as it will be easier to remove UA.)

It runs both of these functions inside of a `DOMContentLoaded` event listener. This is because the `<body>` does not exist yet when `load-analytics.js` code initially runs in the browser, so we have to wait for the DOM to finish loading first before initialising our tracking code otherwise we get errors, because our tracking code tries to add event listeners on the body for things like event tracking. 

`loadUa()` and `loadGa4()` are quite similar. They essentially:

1. Grab the current domain that the code is running on using `window.location.hostname`.
2. Determine the environment you are on by comparing your domain against hardcoded categorised domains.
3. Set the correct environment variables depending on the environment you are on. These env vars are also hardcoded now.
4. Kick off the initialisation of analytics by running their respective `init()` functions, which will then run the `ga4-core.js` code and the UA tracking initialisation code.
5. The `ga4-core.js` / UA code then makes use of these environment variables. It can access them as they are set on the JavaScript `window` object.

I have moved grabbing the current version of publishing components (`gem_version` in the dataLayer) to a meta tag. Now that we can't use `.js.erb`, it's not possible to get the gem version using .js.erb. Therefore I've turned it into a `<meta` tag, and then I grab that using JavaScript. I did consider writing a `<script>` tag in the Ruby template which would be `window.GOVUK.analyticsGa4.vars.gem_version = "<%= GovukPublishingComponents::VERSION %>"` but I thought this was a bit messy writing JavaScript inside the template. Also it would be the first JavaScript that runs on the page, so I'd have to do things like `window.GOVUK.analyticsGa4 = window.GOVUK.analyticsGa4 || {}` to prevent errors with `window.GOVUK` and `analyticsGa4` and `.vars` not yet existing.

### Testing the code

To test this repo, you will need to:
-  checkout the branch `remove-analytics-puppet-dependency` on `static`
- modify the `Gemfile` to use a local version of `govuk_publishing_components` in `static`
-  run `make static && govuk-docker up static-app` 

We need to use `static` to test because the analytics is executed inside the GOVUK `<head>` template, so you need an application which is using this GOVUK `<head>`. So you should be able to see the development GTM script load as well as the `dataLayer` on this page http://static.dev.gov.uk/templates/gem_layout.html.erb  

I'm having a caching issue with my `static` which kept making the code not run properly. If this happens to you then:
- stop static
- run `bundle exec rails assets:precompile` in static
- run `make static && govuk-docker up static-app` again

If you'd like to compare the environment variables  in `govuk-puppet` to check that I've hardcoded the right ones, these are the sources I used: 

- [Production](https://github.com/alphagov/govuk-puppet/blob/main/hieradata_aws/production.yaml#L150)
- [Staging](https://github.com/alphagov/govuk-puppet/blob/main/hieradata_aws/staging.yaml#L223)
- [Integration](https://github.com/alphagov/govuk-puppet/blob/main/hieradata_aws/integration.yaml#L31)
- [Universal Analytics' Production values](https://github.com/alphagov/static/blob/main/config/environments/production.rb#L87)
- [Universal Analytics' "UA-UNSET" values ](https://github.com/alphagov/static/blob/main/config/application.rb#L41)
- The dev GTM environment variables are from Ian. Moving forward we could use this to save time when testing, as our events will now get sent to the a development dashboard. Therefore we can show analysts the dataLayer pushes sent from our local environments, without having to put stuff on Integration first.

## Why

The previous implementation would use `static`. We are moving away from `static`. Also, it would require grabbing environment variables from `govuk-puppet`, and passing them through to JavaScript using a `.js.erb` file. We cannot use `govuk-puppet` anymore due to the new infrastructure and `.js.erb` is a deprecated filetype in Rails.

I am adding `loading-analytics.js` to the GOVUK `<head>`. Most of the other JavaScript files in this repo are initialised in `static` instead. Since the purpose of this PR is to completely remove our reliance on `static`, I'm not going to define the code in this repo and then execute it in `static` otherwise it kind of defeats the purpose of moving the code outside of `static` in the first place.

## Visual Changes

None.
